---
title: "Johns Hopkins Summer Institute: Epigenomics in Public Health Lab 2020"
subtitle: "04 Region Finding, Global, DNAm age, and meQTL Association Analysis"
author: "Kelly Bakulski, David Sosnowski, Shan Andrews, Brion Maher"
date: "May 29, 2020"
output: html_document
---


# Setting up R markdown format
```{r setup, include=FALSE, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Installing relevant packages (only do once per computer)
```{r install packages, include=FALSE, eval=FALSE}
#If you're running code on your personal computer, set eval=TRUE.
#Since we're running code on the Cloud, these packages have already been installed and it is not necessary to re-install. 

### Install BiocManager -- Allows user to install Bioconductor packages
if ( !requireNamespace( "BiocManager" ) ) install.packages( "BiocManager" )

### Install minfi -- 450k analysis package
BiocManager::install( "bumphunter" ) 
BiocManager::install( "minfi" )

### Install RColorBrewer -- Useful package that provides color palettes to help visualize data
install.packages( "RColorBrewer" )

### Install limma -- Functions for single-site association analysis
BiocManager::install( "limma" )

### Install ggplot2 -- Easy plotting functions
install.packages( "ggplot2" )

### Install matrixEQTL --	For fast eQTL and meQTL analyses
install.packages( "MatrixEQTL" )

### Install wateRmelon -- Functions for DNA methylation age Horvath
BiocManager::install( "wateRmelon" )

```

```{r load packages}
#Load the matrixEQTL package
library( MatrixEQTL )
library(ggplot2)
library(minfi)
library(limma)
library(wateRmelon)
```


# Set up color palettes for plotting 
```{r color palettes}
#Set up color palettes for plotting
myColors <- c( "dodgerblue", "firebrick1", "seagreen3" )
graphColors = c( "#023FA5","#7D87B9","#BEC1D4","#D6BCC0","#BB7784", "#D33F6A", "#11C638","#8DD593","#C6DEC7","#EAD3C6",
                "#F0B98D","#EF9708", "#0FCFC0","#9CDED6","#D5EAE7","#F3E1EB","#F6C4E1","#F79CD4", "#4A6FE3","#8595E1",
                "#B5BBE3","#E6AFB9","#E07B91" )
```

# Setting file paths for data, output, scripts
```{r source data/output}
# Setting file paths for data, output, scripts

#The rest of this script assumes that your data are in a folder called "project" on the Cloud.
#It also assumes that your ouptut will be stored in a subfolder called "output" on the Cloud. 
#As you work on your own computer, you will need to specify the folder locations.

# Folder location of the data files
#data_dir <- "/cloud/project/"
data_dir <- "~/Documents/GitHub/Application-and-Analysis-of-Epigenetic-Data-in-Public-Health-Research/data/" #Kelly's personal computer directory
data_dir
# Folder location to put the output files
#output_dir <- "/cloud/project/output/"
output_dir <- "~/Documents/GitHub/Application-and-Analysis-of-Epigenetic-Data-in-Public-Health-Research/output/" #Kelly's personal computer directory
output_dir
# Record the current date for record keeping and appending to output file names
date<-format(Sys.Date(), "%Y%m%d") 
date
```

# Read in the data
```{r load data}
# Need noob, combat.beta, pd.complete, mapped files
load( paste0(data_dir, "noob.rda"))
load(paste0(data_dir,"combat-beta.rda"))
load(paste0(data_dir, "pd-complete.rda"))
load(paste0(data_dir, "mapped-noob.rda"))

#Make sure these output to TRUE
identical( colnames( combat.beta ), colnames( noob ) )
identical( rownames( pd.complete ), colnames( noob ) )
identical( rownames( noob ), rownames( combat.beta ) )

chrnames <- as.character( seqnames( mapped ) )
pos <- as.numeric( start( mapped ) )
```

# Construct regression model
```{r construct regression model}
#Construct the model matrix
mod <- model.matrix( ~ factor( pd.complete$casestatus ) + pd.complete$age + factor( pd.complete$gender ) + factor( pd.complete$smoking ) + pd.complete$PC1 )  
```


# Region finding
```{r region finding analysis}
mybumps <- bumphunter( combat.beta, mod, chr = chrnames, pos = pos,
	pickCutoff = TRUE, pickCutoffQ = 0.975, maxGap = 300, smooth = TRUE,
	smoothFunction = loessByCluster, B = 10, nullMethod = "bootstrap" )

#Annotate the bumphunter output with gene information
mybumps <- mybumps$table
```

# Plot differentially methylated regions
```{r plot differentially methylated regions}

#Plot the top 2 DMRs: here we are formatting the data in a certain way in order to fit
#the ggplot2 plotting function. We have also shifted the position of the controls by 10bp
#so that their methylation distribution is easier to see. 
pdf( paste0(output_dir, "DMRexample.pdf") )  ##############################
for ( i in 1:2 ){
#for (i in 1:1){
	indexstart <- mybumps$indexStart[ i ]
	indexend <- mybumps$indexEnd[ i ]
	grabBeta <- pre.beta[ indexstart:indexend, ]
	formbeta <- c()
	for ( j in 1:nrow( grabBeta ) ){
		tempbeta <- grabBeta[ j, ]
		formbeta <- c( formbeta, tempbeta )
	}
	reppos <- rep( pos[ indexstart:indexend ], each = ncol( pre.beta ) )
	#reppos.real<-pos[reppos]
	xmin <- min( reppos ) - 100
	xmax <- max( reppos ) + 100
	status <- rep( pd$casestatus, nrow( grabBeta ) )
	reppos.real.shift <- ifelse( status == "Control", reppos + 10, reppos )
	toplot <- data.frame( ( formbeta ), reppos.real.shift, status )
	rownames( toplot )<-c( )
	colnames( toplot )<-c("Beta","Position","Status")
	#toplot$Status<-ifelse(toplot$Status==2,"normal","autism")
	p2 <- ggplot( data= toplot, aes( x = ( Position ),y = ( Beta ),color = factor( Status ) ) ) + geom_point( size = 0.75 ) + scale_colour_manual(values = c( "dodgerblue","black" ) ) +
	theme( legend.direction = "horizontal", legend.position = c( 0.5,0.95 ), legend.title = element_blank( ), panel.background = element_blank( ) ) +
	ylab( "Percent Methylation" ) + xlab( "Position" ) +
	scale_x_continuous( limits = c( xmin, xmax ) ) + stat_smooth( method = "loess", se = FALSE ) +
		scale_y_continuous( limits = c( 0,1 ), breaks = seq( 0, 1, by = 0.25 ), labels = c( "0","25","50","75","100" ) )
		#+stat_summary(fun.y=mean,geom="line",size=1)
	print( p2 )
}
dev.off()
```


# DNAm Age analyses
# Horvath DNAm clock calculation
```{r Horvath DNAm age analyses}
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6075434/
horvath <- read.csv(paste0(data_dir,"HorvathSkinBloodCoefs.csv"))
horvath.coef <- horvath$Coef
names(horvath.coef) <- horvath$ID
table(names(horvath.coef)[-1] %in% rownames(combat.beta))

pd.complete$horvath.age <- agep(combat.beta, coef=horvath.coef, method='horvath')
head(pd.complete$horvath.age)

pdf(file=paste0(output_dir, "Horvath_DNA_methylation_Age.pdf"))
plot(pd.complete$age, pd.complete$horvath.age, xlab="Chronological Age", ylab="Horvath DNA methylation age", las=1)
abline(lm(pd.complete$horvath.age~ pd.complete$age), col="grey")
palette(myColors)
plot(pd.complete$age, pd.complete$horvath.age, xlab="Chronological Age", ylab="Horvath DNA methylation age", las=1, col=as.factor(pd.complete$casestatus))
abline(lm(pd.complete$horvath.age~ pd.complete$age, subset=pd.complete$casestatus=="Control"), col=myColors[1])
abline(lm(pd.complete$horvath.age~ pd.complete$age, subset=pd.complete$casestatus=="RA"), col=myColors[2])
legend("topleft", col=myColors[1:2], legend=c("Control", "RA"), pch=21)
dev.off()
```

# Levine Phenotypic DNAm clock calculation
```{r Levine phenotypic DNAm age analyses}
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5940111/
levine <- read.csv(paste0(data_dir,"LevineCoefs.csv"))
levine.coef <- levine$Weight
names(levine.coef) <- levine$CpG

table(names(levine.coef)[-1] %in% rownames(combat.beta))

betas.lev <- data.frame(matrix(nrow=length(levine.coef), ncol=ncol(combat.beta)))
betas.lev[2:length(levine.coef),] <- combat.beta[rownames(combat.beta) %in% names(levine.coef),]
betas.lev[1:5,1:5]
rownames(betas.lev)[1] <- 'Intercept'
betas.lev[1,] <- 1
betas.lev[1:5,1:5]

pd.complete$levine.age <- colSums(betas.lev * levine.coef)
head(pd.complete$levine.age)

pdf(file=paste0(output_dir, "Levine_DNA_methylation_Age.pdf"))
plot(pd.complete$age, pd.complete$levine.age, xlab="Chronological Age", ylab="Levine DNA methylation age", las=1)
abline(lm(pd.complete$levine.age~ pd.complete$age), col="grey")

palette(myColors)
plot(pd.complete$age, pd.complete$levine.age, xlab="Chronological Age", ylab="Levine DNA methylation age", las=1, col=as.factor(pd.complete$casestatus))
abline(lm(pd.complete$levine.age~ pd.complete$age, subset=pd.complete$casestatus=="Control"), col=myColors[1])
abline(lm(pd.complete$levine.age~ pd.complete$age, subset=pd.complete$casestatus=="RA"), col=myColors[2])
legend("topleft", col=myColors[1:2], legend=c("Control", "RA"), pch=21)
dev.off()
```


# meQTL analyses

```{r load genotype data}
#Load the genotype object
load( paste0(data_dir, "genotypes.rda" ))
load( paste0(data_dir, "snp.pos.rda" ))
```

```{r prepare genetic and methylation data}

#Subset methylation data and genomic positions to only our 
#chromosome of interest
B.mychr <- combat.beta[ which( chrnames == "chr22" ), ]
pos.mychr <- pos[ which( chrnames == "chr22" ) ]

#Subset the methylation data to probes in this particular region
B.LDblock <- B.mychr[ which( pos.mychr > 17583446 & pos.mychr < 17666446 ), ]   
pos.LDblock <- pos.mychr[ which( pos.mychr > 17583446 & pos.mychr < 17666446 ) ]

#Let's be sure to match to the samples in our methylation matrix
genotypes <- genotypes[, match (colnames( B.LDblock ), colnames( genotypes ) ) ]

#Format the genotypes and methylation objects for the package
genotypes.format <- SlicedData$new( genotypes )
meth.format <- SlicedData$new( B.LDblock )
```

# Calculate meQTLs
```{r}
#Call the function
results <- Matrix_eQTL_main( genotypes.format, meth.format, pvOutputThreshold = 0.05,
	snpspos = snp.pos,genepos = pos.LDblock, output_file_name = NULL, output_file_name.cis = NULL )

#Grab the meQTL results in a separate object for ease of plotting downstream
results.table <- results$all$eqtls

#Let's examine the spatial relationship between SNPs and CpG sites.
#First we make sure we grab the right positions
results.table$SNPpos <- snp.pos[ match( results.table$snps, rownames( genotypes ) ) ]
results.table$CGpos <- pos.LDblock[ match( results.table$gene, rownames( B.LDblock ) ) ]

#Transform the p-value to the -log10 scale
results.table$transP <- ( -1*log( results.table$pvalue, base = 10 ) )

results.table$Distance <- results.table$CGpos - results.table$SNPpos

```

# Plot meQTL
```{r plot meQTL}
pdf( paste0(output_dir, "Volcano_meQTL.pdf" ))
with( results.table, plot( Distance, transP, pch = 20, main = NULL, 
	xlab = "CpG Position - SNP Position (Kb)", ylab = "-log10 p-value" ) )
dev.off()
	
```



# ADDENDUM: READ DATA FROM GEO

```{r prepare for data download from GEO, eval=FALSE}
#Use the 'GEOquery' package
library( GEOquery )
# set your data directory to where you want to download the files
data_dir <- "/cloud/project/data/"
data_dir
```


# Download the data from GEO
```{r download the supplementary files from geo, eval=FALSE}

#Download the supplementary files attached to this 
#GEO ID. Raw .idat files are part of this group of
#supplementary files. 
getGEOSuppFiles( "GSE42861", makeDirectory = TRUE, baseDir = data_dir )

#General data available for this GEOID
mystudy <- getGEO( GEO = "GSE42861", destdir = getwd( ) )

#Phenotype/covariate information for these samples  
mypheno <- ( phenoData( mystudy$GSE42861_series_matrix.txt.gz ) )
variables <- varMetadata( mypheno )
```
